packages/styled-components/src/utils/flatten.js

```diff
@@ -55,23 +55,19 @@ export default function flatten(chunk: any, executionContext: ?Object, styleShee
 
   /* Either execute or defer the function */
   if (isFunction(chunk)) {
-    if (executionContext) {
-      let shouldThrow = false;
+    if (isStatelessFunction(chunk) && executionContext) {
+      const result = chunk(executionContext);
 
-      try {
-        // eslint-disable-next-line new-cap
-        if (isElement(new chunk(executionContext))) {
-          shouldThrow = true;
-        }
-      } catch (e) {
-        /* */
+      if (process.env.NODE_ENV !== 'production' && isElement(result)) {
+        // eslint-disable-next-line no-console
+        console.warn(
+          `${getComponentName(
+            chunk
+          )} is not a styled component and cannot be referred to via component selector. See https://www.styled-components.com/docs/advanced#referring-to-other-components for more details.`
+        );
       }
 
-      if (shouldThrow) {
-        throw new StyledError(13, getComponentName(chunk));
-      }
-
-      return flatten(chunk(executionContext), executionContext, styleSheet);
+      return flatten(result, executionContext, styleSheet);
     } else return chunk;
```

Comment:
The else branch here should also have the warning, no? If `isStatelessFunction(chunk)` returns false that implies a class, which is not a valid thing to interpolate.

Reply:
ping @alansouzati

Reply:
sorry I've missed this. I think the warning should be only in one place. probably by moving things around. let me take a careful pass at this.

Reply:
I don't think we should add a warning in the else.

`isStatelessFunction(chunk) && executionContext`

this is deferring the execution since `executionContext` is empty.

Reply:
@mxstbr please advise here. I don't have the full context of this flatten operation so I can definitely be missing something.

Reply:
Ok I see! Yes this is perfect, thanks @alansouzati!